heron/packing/src/java/com/twitter/heron/packing/builder/HomogeneityScorer.java

```diff
@@ -0,0 +1,51 @@
+// Copyright 2016 Twitter. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.twitter.heron.packing.builder;
+
+import com.twitter.heron.spi.packing.PackingPlan;
+
+/**
+ * Scores based on how homogeneous a container is relative to a given component, from highest to
+ * lowest. So a container with all the same components (e.g., AAAA) will get the highest score for
+ * A, while a container with a mix of many components (e.g. ABCD) will get a low score.
+ */
+public class HomogeneityScorer implements Scorer<Container> {
+  private String componentName;
+
+  public HomogeneityScorer(String componentName) {
+    this.componentName = componentName;
+  }
+
+  @Override
+  public boolean sortAscending() {
+    return false;
+  }
+
+  @Override
+  public double getScore(Container container) {
+    int totalComponentInstances = 0;
+    int totalInstances = container.getInstances().size();
+
+    if (totalInstances == 0) {
+      return 0;
+    }
+
+    for (PackingPlan.InstancePlan instancePlan : container.getInstances()) {
+      if (componentName.equals(instancePlan.getComponentName())) {
+        totalComponentInstances++;
+      }
+    }
+    return (double) totalComponentInstances / totalInstances;
```

Comment:
I think you lose opportunities for removing containers with this scoring function. Let's take a container BBCD and one ABC and you want to remove 1 B. It is better to remove it from ABC than BBCD because you maximize the probability to remove this container by doing that. The current scoring function will remove from BBCD. I think you need to categorize containers and use total number of instances as the scoring function within each category. I have another example in the discussion thread. In general optimizing for imbalance will not let you remove containers easily. You can cover all cases by using a two-step approach as discussed in #1560